File: packages/agent/src/infrastructure/events/types/rules.ts
/**
 * @fileoverview Rules Events
 *
 * Events for rules file loading.
 */

import type { BaseEvent } from './base.js';

// =============================================================================
// Rules Events
// =============================================================================

/** Level of a rules file in the hierarchy */
export type RulesLevel = 'global' | 'project' | 'directory';

/** Information about a single rules file */
export interface RulesFileInfo {
  /** Absolute path to the file */
  path: string;
  /** Path relative to working directory (or absolute if outside) */
  relativePath: string;
  /** Level in the hierarchy */
  level: RulesLevel;
  /** Depth from project root (0 = root, -1 = global) */
  depth: number;
  /** File size in bytes */
  sizeBytes: number;
}

/**
 * Payload for rules.loaded event
 * Emitted once per session when rules files are loaded
 */
export interface RulesLoadedPayload {
  /** List of loaded rules files */
  files: RulesFileInfo[];
  /** Total number of rules files loaded */
  totalFiles: number;
  /** Estimated token count for merged rules content */
  mergedTokens: number;
  /** Number of additional dynamic rules discovered (subfolder CLAUDE.md/AGENTS.md) */
  dynamicRulesCount?: number;
}

/**
 * Rules loaded event - emitted at session start when rules files are detected
 */
export interface RulesLoadedEvent extends BaseEvent {
  type: 'rules.loaded';
  payload: RulesLoadedPayload;
}

// =============================================================================
// Rules Indexed Events (scoped CLAUDE.md/AGENTS.md files)
// =============================================================================

/** Payload for rules.indexed event — emitted at session start after discovery */
export interface RulesIndexedPayload {
  totalRules: number;
  globalRules: number;
  scopedRules: number;
  files: Array<{
    relativePath: string;
    isGlobal: boolean;
    scopeDir: string;
    sizeBytes: number;
  }>;
}

/** Rules indexed event - emitted when scoped rules are discovered and indexed */
export interface RulesIndexedEvent extends BaseEvent {
  type: 'rules.indexed';
  payload: RulesIndexedPayload;
}

File: packages/agent-rs/crates/tron-events/src/types/payloads/rules.rs
//! Rules event payloads.

use serde::{Deserialize, Serialize};

/// Payload for `rules.loaded` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulesLoadedPayload {
    /// Rules files loaded.
    pub files: Vec<RulesFileInfo>,
    /// Total files loaded.
    pub total_files: i64,
    /// Total merged token count.
    pub merged_tokens: i64,
    /// Number of dynamic rules.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dynamic_rules_count: Option<i64>,
}

/// Info about a loaded rules file.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulesFileInfo {
    /// Absolute path.
    pub path: String,
    /// Relative path.
    pub relative_path: String,
    /// Rules level: global, project, directory.
    pub level: String,
    /// Directory depth.
    pub depth: i64,
    /// File size in bytes.
    pub size_bytes: i64,
}

/// Payload for `rules.indexed` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulesIndexedPayload {
    /// Total number of rules.
    pub total_rules: i64,
    /// Number of global rules.
    pub global_rules: i64,
    /// Number of scoped rules.
    pub scoped_rules: i64,
    /// Files indexed.
    pub files: Vec<RulesIndexedFile>,
}

/// Info about an indexed rules file.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RulesIndexedFile {
    /// Relative path.
    pub relative_path: String,
    /// Whether it's a global rules file.
    pub is_global: bool,
    /// Scope directory.
    pub scope_dir: String,
    /// File size in bytes.
    pub size_bytes: i64,
}
